File: iOS_JavaScript/iOS-callJavaScript/OC&JavaScript/JavaScriptCore-OCExecuteJS/ViewController.m
//
//  ViewController.m
//  JavaScriptCore-OCExecuteJS
//
//  Created by 孟跃平 on 2017/6/18.
//  Copyright © 2017年 www.mengyueping.com. All rights reserved.
//
/*
 * 使用JavaScriptCore执行一段JavaScript代码
 *
 * 使用OC的Block定义js函数。
 */

#import "ViewController.h"
#import <JavaScriptCore/JavaScriptCore.h>

@interface ViewController ()

@end
@implementation ViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    
#pragma mark 初始化 JSContext 对象
    
    JSContext *context = [[JSContext alloc] init];
    // 异常处理回调
    context.exceptionHandler = ^(JSContext *context, JSValue *exception) {
        NSLog(@"%@", exception);
        context.exception = exception;
        /*
         此处打印js异常错误，JSContext不会主动抛出js异常。
         常见异常：
             ReferenceError: Can't find variable:
             TypeError: undefined is not an object
         */
    };
    
#pragma mark 执行js脚本，返回执行结果
    
    NSString *jsStr1 = @"1+2";
    JSValue *value1 = [context evaluateScript:jsStr1]; // 返回的JSValue对象，存储的是js计算的返回结果。不存储定义的变量。
    NSLog(@"value1 JS: %@  -> OC: %@",value1, value1.toNumber);
    
#pragma mark 执行js脚本，存储js脚本中定义的变量到 JSContext 对象，并key-value方法取出
    
    NSString *jsStr2 = @"var a = 1; var b = 2;";
    JSValue *value2 = [context evaluateScript:jsStr2];
    JSValue *jsValueA = context[@"a"];  // 运行js，定义的变量，存储在context中。
    JSValue *jsValueB = context[@"b"];
    NSLog(@"value2 JS: %@  -> OC: %@",value2, value2.toObject); // value2 JS: undefined  -> OC: (null)
    NSLog(@"JS a: %@;  JS a: %@", jsValueA, jsValueB); // JS a: 1;  JS a: 2
    
#pragma mark 使用key-value方法“取出”并“修改”js对象集合类型的元素值
    
    NSString *jsStr3 = @"var arr = [88, 'mengyueping', 66];";
    JSValue *value3 = [context evaluateScript:jsStr3];
    JSValue *jsArr = context[@"arr"];
    NSLog(@"value3 JS: %@  -> OC: %@",value3, value3.toObject); // value2 JS: undefined  -> OC: (null)
    NSLog(@"JS Array: %@;  Length: %@; jsArr[0]：%@", jsArr, jsArr[@"length"], jsArr[0]); // JS Array: 20,10,www.mengyueping.com;  Length: 3; jsArr[0]：20
    jsArr[0] = @"www.";
    jsArr[2] = @".com";
    NSLog(@"通过下标对js取值赋值：JS Array: %@;  Length: %@; jsArr[0]：%@", jsArr, jsArr[@"length"], jsArr[0]); // JS Array: 20,10,www.mengyueping.com;  Length: 3; jsArr[0]：20
    
#pragma mark 执行js脚本，存储的js集合对象，转换为OC数组对象
    
    NSArray *ocArr = jsArr.toArray;
    NSLog(@"js Arr -> OC Arr: %@", ocArr);
    /*
     js Arr -> OC Arr: (
     "www.",
     mengyueping,
     ".com"
     )
     */
    
#pragma mark 赋值js集合对象的值，没有此元素自动添加补齐
    
    jsArr[8] = @8;
    NSLog(@"通过下标对js取值赋值：JS Array: %@;  Length: %@; jsArr[0]：%@", jsArr, jsArr[@"length"], jsArr[0]);
    // 通过下标对js取值赋值：JS Array: www.,mengyueping,.com,,,,,,8;  Length: 9; jsArr[0]：www.
    NSLog(@"js Arr -> OC Arr: %@", jsArr.toArray);
    /*
     js Arr -> OC Arr: (
     "www.",
     mengyueping,
     ".com",
     "<null>",
     "<null>",
     "<null>",
     "<null>",
     "<null>",
     8
     )
     */
    
#pragma mark 使用js语法创建函数，并执行js，把函数存储到JSContext 对象, 调用js函数传参
    
    [context evaluateScript:@"function add(a, b){ return a + b; }"];
    JSValue *addValue = context[@"add"]; // js函数
    NSLog(@"Func: %@", addValue);  // Func: function add(a, b){ return a + b; }
    
    // 取出js函数，调用函数 - (JSValue *)callWithArguments:(NSArray *)arguments;
    JSValue *sum = [addValue callWithArguments:@[@1,@2]];
    NSLog(@"Sum: %d", sum.toInt32); // Sum: 3
    
    
#pragma mark 另一种简单的调用js函数的方法
    
    JSValue *jsValue = [context evaluateScript:@"function multiply(a, b){ return a * b; }"];
    JSValue *multiplyValue = [jsValue.context.globalObject invokeMethod:@"multiply" withArguments:@[@3,@6]];
//    JSValue *multiplyValue = [context.globalObject invokeMethod:@"multiply" withArguments:@[@3,@6]];
    NSLog(@"multiplyValue: %d",multiplyValue.toInt32); // multiplyValue: 18
    
#pragma mark 使用OC中的Block定义js函数，并存储到JSContext 对象
    
    context[@"log"] = ^(){
        NSLog(@"++++++Begin Log++++++");
        
        NSArray *args = [JSContext currentArguments];
        for (JSValue *jsVal in args) {
            NSLog(@"%@",jsVal);
        }
        
        JSValue *this = [JSContext currentThis];
        NSLog(@"this: %@", this);
        
        NSLog(@"---End Log------");
    };
    // 执行js，调用使用Block自定义的js函数
    [context evaluateScript:@"log('mengyueping', [10,20], {'hello': 'world', 'number': '100'})"];
    /*
     ++++++Begin Log++++++
     mengyueping
     10,20
     [object Object]
     this: [object GlobalObject]
     ---End Log------
     */
}

@end

File: iOS_JavaScript/iOS-callJavaScript/Swift&JavaScript/JavaScriptCore-SwiftExecuteJS/ViewController.swift
//
//  ViewController.swift
//  JavaScriptCore-SwiftExecuteJS
//
//  Created by 孟跃平 on 2017/6/19.
//  Copyright © 2017年 www.mengyueping.com. All rights reserved.
//

/*
 * 使用JavaScriptCore执行一段JavaScript代码
 *
 * 使用OC的Block定义js函数。
 */

import UIKit
import JavaScriptCore

class ViewController: UIViewController {
    
    lazy var context = JSContext()!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // MARK: 初始化JSContext对象及异常处理
        context.exceptionHandler = { (context, exception) in
            guard let excep = exception else {
                return
            }
            print(excep)
            /*
             此处打印js异常错误，JSContext不会主动抛出js异常。
             常见异常：
             ReferenceError: Can't find variable:
             TypeError: undefined is not an object
             */
            context?.exception = excep
        }
        
        // MARK: 执行带返回结果js脚本
        let jsStr1 = "1+2"
        let jsvalue1 = context.evaluateScript(jsStr1)
        print("value1 JS: \(String(describing: jsvalue1))  -> Swift: \(String(describing: jsvalue1?.toNumber()))")
        // value1 JS: Optional(3)  -> Swift: Optional(3)
        
        
        // MARK: 执行不带返回结果js脚本
        let jsStr2 = "var a = 1; var b = 2;"
        let jsvalue2 = context.evaluateScript(jsStr2)
        print("value2 JS: \(String(describing: jsvalue2))  -> Swift: \(String(describing: jsvalue2?.toObject))");
        //value2 JS: Optional(undefined)  -> Swift: Optional((Function))
        
        
        // MARK: 取出js脚本执行后存储在JSContext对象中的变量
        let jsValueA = context.objectForKeyedSubscript("a")
        let jsValueB = context.objectForKeyedSubscript("b")
        NSLog("JS a: \(String(describing: jsValueA));  JS b: \(String(describing: jsValueB))") //  JS a: Optional(1);  JS b: Optional(2)
        
        
        // MARK: 取出存储的js变量，并修改
        let jsStr3 = "var arr = [88, 'mengyueping', 66];"
        let value3 = context.evaluateScript(jsStr3)
        let jsArr = context.objectForKeyedSubscript("arr")
        print("value3 JS: \(String(describing: value3))  -> Swift: \(String(describing: value3?.toObject))");
        // value3 JS: Optional(undefined)  -> Swift: Optional((Function))
        print("JS Array: \(String(describing: jsArr));  Length: \(String(describing: jsArr?.objectForKeyedSubscript("length"))); jsArr[0]：\(String(describing: jsArr?.objectAtIndexedSubscript(0)))" )
        // JS Array: Optional(88,mengyueping,66);  Length: Optional(3); jsArr[0]：Optional(88)
        jsArr?.setValue("www.", at: 0)
        jsArr?.setValue(".com", at: 2)
        print("通过下标对js取值赋值：JS Array: \(String(describing: jsArr));  Length: \(String(describing: jsArr?.objectForKeyedSubscript("length"))); jsArr[0]：\(String(describing: jsArr?.objectAtIndexedSubscript(0)))" )
        // 通过下标对js取值赋值：JS Array: Optional(www.,mengyueping,.com);  Length: Optional(3); jsArr[0]：Optional(www.)
        
        
        // MARK: 取出存储的js集合对象，并转为Swift数组对象
        let swiftArr = jsArr?.toArray()
        print("js Arr -> Swift Arr: \(String(describing: swiftArr))");
        /*
         js Arr -> Swift Arr: Optional([www., mengyueping, .com])
         */
        
        
        // MARK: 取出存储的js集合对象，并直接使用Swift对象给js对象赋值
        jsArr?.setValue(8, at: 8)
        print("通过下标对js取值赋值：JS Array: \(String(describing: jsArr));  Length: \(String(describing: jsArr?.objectForKeyedSubscript("length"))); jsArr[0]：\(String(describing: jsArr?.objectAtIndexedSubscript(0)))" )
        // 通过下标对js取值赋值：JS Array: Optional(www.,mengyueping,.com,,,,,,8);  Length: Optional(9); jsArr[0]：Optional(www.)
        print("js Arr -> Swift Arr: \(String(describing: jsArr?.toArray()))")
        /*
         js Arr -> Swift Arr: Optional([www., mengyueping, .com, <null>, <null>, <null>, <null>, <null>, 8])
         */
        
        
        // MARK: 取出存储的js函数，并执行
        context.evaluateScript("function add(a, b){ return a + b; }")
        let addValue = context.objectForKeyedSubscript("add") // js函数
        print("Func: \(String(describing: addValue))")  // Func: Optional(function add(a, b){ return a + b; })
        // 取出js函数，调用函数 open func call(withArguments arguments: [Any]!) -> JSValue!
        let sum = addValue?.call(withArguments:[1,2])
        print("Sum: \(String(describing: sum?.toInt32()))") // Sum: Optional(3)
        
        
        // MARK:  调用js函数的另一种简单方法
        let jsValue = context.evaluateScript("function multiply(a, b){ return a * b; }")
        let multiplyValue = jsValue?.context.globalObject.invokeMethod("multiply", withArguments: [3,6])
//        let multiplyValue = context.globalObject.invokeMethod("multiply", withArguments: [3,6])
        print("multiplyValue: \(String(describing: multiplyValue?.toInt32()))"); // multiplyValue: Optional(18)
        
        
        // MARK: 把OC中Closure转换成JS函数，并存储到JSContext对象
        let block: @convention(block) () -> () = {
            print("++++++Begin Log++++++")
            
            let args = JSContext.currentArguments()
            
            for jsVal in args! {
                print(jsVal)
            }
            
            let this = JSContext.currentThis()
            print("this: \(String(describing: this))")
            
            print("---End Log------")
        }
        context.setObject(block, forKeyedSubscript: NSString(string: "log"))
        // 执行js，调用使用Block自定义的js函数
        context.evaluateScript("log('mengyueping', [10,20], {'hello': 'world', 'number': '100'})")
        /*
         ++++++Begin Log++++++
         mengyueping
         10,20
         [object Object]
         this: Optional([object GlobalObject])
         ---End Log------
         */
        
    }
    

}

